### ISSUE TITLE
Diagnostic Parser (problem matcher): relative paths doesn't work - VSCode.URI doesn't support it

### ISSUE THREAD

**Issue by FoxDevilsWild** [microsoft/vscode-cmake-tools#299]
### Brief Issue Summary
After building my project I'm using the problem matcher to "jump-to-error" in the problem window. Apparently he can't find the desired file because its path is wrong. The reason lies probably in the relative path output of the build process, but lets look at the problem with an example:

### Platform and Versions

- **Operating System**: Windows 10 Enterprise
- **CMake Version**: 2.8.12
- **VSCode Version**: 1.19.1
- **CMake Tools Extension Version**: 0.10.4
- **Compiler/Toolchain**: Visual C++ 2010

### Example

I will show you the paths/output of the different Visual Studio Code Windows. The error will be matched (extracted) correctly but the given path to the file will be wrong.

#### Output Window

`2>..\..\..\src\DataRecorder\DataRecorder.cpp(78): error C2143: Syntaxfehler: Es fehlt ';' vor '}'`

#### Absolute path of that file

`c:\daten\04_workspace\HoD\src\DataRecorder\DataRecorder.cpp`

#### Path in Problem Window 

`c:\daten\src\DataRecorder\DataRecorder.cpp`

#### Build directory

`"${workspaceRoot}/build"` which is `c:\daten\04_workspace\HoD\build`

As you can see Visual Studio Code expects the file at the wrong place. The shown paths suggests that the used path in the problem window was generated by the build directory and the given path of the output window. (3 folders up from `build` - we are in `c:\daten`)

I actually don't know if this is a bug or i didn't configure the project the right way. I got it to run with a Visual Studio Code task:

```
    "version": "2.0.0",
    "tasks": [
        {
            "label": "CMake Build - RelWithDebInfo",
            "type": "shell",
            "command": "cmake --build c:/Daten/04_Workspace/HoD/build --target DataRecorder --config RelWithDebInfo",
            "presentation": {
                "reveal": "always",
                "echo": true
            },
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "problemMatcher":{
                "owner": "cpp",
                "fileLocation": "relative",
                "pattern":[
                    {
                        "regexp": "([\\w+|\\\\]*\\.\\w+)\\((\\d+)\\)\\: (warning|error) (.*)$",
                        "file": 1,
                        "location": 2,
                        "severity": 3,
                        "message": 4
                    }
                ]
            }
        }
    ]
}
```
There i was able to select the usage of _relative paths_. I didn't find such possibility in the vscode-cmake-tools settings.

### Probable Reason

The problem matcher/ diagnostic parser is found in the the `cmake.js`. I think the problem lies in the line 450 (version 0.10.4):

```
diagset.set(vscode.Uri.file(filepath), [...diags.values()]);
```

According to this issue [#34449](https://github.com/Microsoft/vscode/issues/34449) of Visual Studio Code the URI API can't handle relative paths.

I will try to investigate further.

### Full output of Output Window

```
Loading CMake Tools from c:\Daten\04_Workspace\HoD\.vscode\.cmaketools.json
Execute cmake with arguments: -Cc:\daten\04_workspace\hod\build\CMakeTools\InitializeCache.cmake,-Hc:/daten/04_workspace/hod,-Bc:/daten/04_workspace/hod/build
[vscode] Executing command: cmake -Cc:\daten\04_workspace\hod\build\CMakeTools\InitializeCache.cmake -Hc:/daten/04_workspace/hod -Bc:/daten/04_workspace/hod/build
loading initial cache file c:\daten\04_workspace\hod\build\CMakeTools\InitializeCache.cmake
cxx Flags: /DWIN32 /D_WINDOWS /W3 /GR /EHsc /W4
-- Detected platform is: win64_vc100
Qt: Untested Windows version 6.2 detected!
Qt: Untested Windows version 6.2 detected!
Qt: Untested Windows version 6.2 detected!
-- Found Qt4: c:/Daten/02_Libraries/01_Develoment/qt-4.7.1/bin/qmake (found version "4.7.1") 
Qt: Untested Windows version 6.2 detected!
Qt: Untested Windows version 6.2 detected!
Qt: Untested Windows version 6.2 detected!
-- Found Qt4: c:/Daten/02_Libraries/01_Develoment/qt-4.7.1/bin/qmake (found version "4.7.1") 
-- Configuring done
-- Generating done
-- Build files have been written to: c:/Daten/04_Workspace/HoD/build
[vscode] cmake exited with return code 0
Execute cmake with arguments: --build,c:/daten/04_workspace/hod/build,--target,ALL_BUILD,--config,RelWithDebInfo,--
[vscode] Executing command: cmake --build c:/daten/04_workspace/hod/build --target ALL_BUILD --config RelWithDebInfo --

Microsoft (R) Visual Studio Version 10.0.40219.1.
Copyright (C) Microsoft Corp. Alle Rechte vorbehalten.
1>------ Erstellen gestartet: Projekt: ZERO_CHECK, Konfiguration: RelWithDebInfo x64 ------
1>Der Buildvorgang wurde am 10.01.2018 14:42:06 gestartet.
1>InitializeBuildStatus:
1>  "x64\RelWithDebInfo\ZERO_CHECK\ZERO_CHECK.unsuccessfulbuild" wird erstellt, da "AlwaysCreate" angegeben wurde.
1>FinalizeBuildStatus:
1>  Die Datei "x64\RelWithDebInfo\ZERO_CHECK\ZERO_CHECK.unsuccessfulbuild" wird gel�scht.
1>  Aktualisieren des Timestamps von "x64\RelWithDebInfo\ZERO_CHECK\ZERO_CHECK.lastbuildstate".
1>
1>Das Erstellen war erfolgreich.
1>
1>Verstrichene Zeit 00:00:00.11
2>------ Erstellen gestartet: Projekt: DataRecorder, Konfiguration: RelWithDebInfo x64 ------
2>Der Buildvorgang wurde am 10.01.2018 14:42:06 gestartet.
2>InitializeBuildStatus:
2>  Aktualisieren des Timestamps von "DataRecorder.dir\RelWithDebInfo\DataRecorder.unsuccessfulbuild".
2>CustomBuild:
2>  Alle Ausgaben sind aktuell.
2>ClCompile:
2>  DataRecorder.cpp
2>  Statische Bindung des ADTF Plugin-SDK
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(339): warning C4100: '__exception_ptr': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(338): warning C4100: 'ppvObject': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(337): warning C4100: 'pstrInstanceInfo': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(336): warning C4100: 'strObjectId': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(335): warning C4100: 'nCmdId': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(357): warning C4100: '__exception_ptr': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(357): warning C4100: 'ppvObject': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(357): warning C4100: 'pstrInstanceInfo': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(357): warning C4100: 'strObjectId': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(357): warning C4100: 'nCmdId': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(375): warning C4100: '__exception_ptr': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(375): warning C4100: 'ppvObject': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(375): warning C4100: 'pstrInstanceInfo': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(375): warning C4100: 'strObjectId': Unreferenzierter formaler Parameter
2>c:\daten\02_libraries\00_adtf\2.14.0\include\ucom\plugin.h(375): warning C4100: 'nCmdId': Unreferenzierter formaler Parameter
2>..\..\..\src\DataRecorder\DataRecorder.cpp(23): warning C4100: 'lpReserved': Unreferenzierter formaler Parameter
2>..\..\..\src\DataRecorder\DataRecorder.cpp(23): warning C4100: 'ul_reason_for_call': Unreferenzierter formaler Parameter
2>..\..\..\src\DataRecorder\DataRecorder.cpp(23): warning C4100: '__exception_ptr': Unreferenzierter formaler Parameter
2>..\..\..\src\DataRecorder\DataRecorder.cpp(23): warning C4100: '__exception_ptr': Unreferenzierter formaler Parameter
2>..\..\..\src\DataRecorder\DataRecorder.cpp(78): error C2143: Syntaxfehler: Es fehlt ';' vor '}'
2>
2>Fehler beim Erstellen
2>
2>Verstrichene Zeit 00:00:01.37
3>------ Erstellen gestartet: Projekt: ALL_BUILD, Konfiguration: RelWithDebInfo x64 ------
3>Der Buildvorgang wurde am 10.01.2018 14:42:08 gestartet.
3>InitializeBuildStatus:
3>  "x64\RelWithDebInfo\ALL_BUILD\ALL_BUILD.unsuccessfulbuild" wird erstellt, da "AlwaysCreate" angegeben wurde.
3>CustomBuild:
3>  Build all projects
3>FinalizeBuildStatus:
3>  Die Datei "x64\RelWithDebInfo\ALL_BUILD\ALL_BUILD.unsuccessfulbuild" wird gel�scht.
3>  Aktualisieren des Timestamps von "x64\RelWithDebInfo\ALL_BUILD\ALL_BUILD.lastbuildstate".
3>
3>Das Erstellen war erfolgreich.
3>
3>Verstrichene Zeit 00:00:00.13
```

**Comment by vector-of-bool** [comment]
Sorry for the delay. I'll take a look into the path handling and see what might be happening.

**Comment by srinivasnks** [comment]
I am not sure if the issue is related but I saw a similar issue with golang and symlinks. I opened an issue with https://github.com/Microsoft/vscode (see here: https://github.com/Microsoft/vscode/issues/43071)

What I am noticing is that vscode is resolving the symlinks differently for errors vs warnings. For errors, it is skipping a folder from the path (similar to the issue mentioned above) but for warnings, it is following the symlink just fine.
